batch-stark/src/common.rs

```diff
@@ -198,6 +221,53 @@ where
             })
         };
 
-        Self { preprocessed }
+        let lookups: Vec<Vec<Lookup<Val<SC>>>> =
+            airs.iter_mut().map(|air| air.get_lookups()).collect();
+
+        Self {
+            preprocessed,
+            lookups,
+        }
     }
 }
+
+pub(crate) fn get_perm_challenges<SC: SGC, LG: LookupGadget>(
+    challenger: &mut SC::Challenger,
+    all_lookups: &[Vec<Lookup<Val<SC>>>],
+    lookup_gadget: &LG,
+) -> Vec<Vec<SC::Challenge>> {
+    let num_challenges_per_lookup = lookup_gadget.num_challenges();
+    let mut global_perm_challenges = HashMap::new();
+
+    all_lookups
+        .iter()
+        .map(|contexts| {
+            // Pre-allocate for the instance's challenges.
+            let num_challenges = contexts.len() * num_challenges_per_lookup;
+            let mut instance_challenges = Vec::with_capacity(num_challenges);
+
+            for context in contexts {
+                match &context.kind {
+                    Kind::Global(name) => {
+                        // Get or create the global challenges.
+                        let cs: &mut Vec<SC::Challenge> =
+                            global_perm_challenges.entry(name).or_insert_with(|| {
+                                (0..num_challenges_per_lookup)
+                                    .map(|_| challenger.sample_algebra_element())
+                                    .collect()
+                            });
+                        // Extend from the slice directly without vector cloning
+                        instance_challenges.extend(cs.iter().copied());
```

Comment:
Can't we do something like this?

```suggestion
                        let challenges = cache.entry(name.clone()).or_insert_with(|| {
                            (0..n)
                                .map(|_| challenger.sample_algebra_element())
                                .collect()
                        });
                        // Maybe here we can copy without iterator                        instance_challenges.extend_from_slice(challenges);
```